step 2 - first commit

andrey340 opened the pull request:
For check of sprint 1 - step 2

Thanks for checking my code
With regards Andrey Volobuev

gennady-bars commented on src/components/hocs/withModal.jsx:
```diff
@@ -0,0 +1,46 @@
+const withModal = (WrappedComponent) => (props) => {
+    const {modalType = 'error', modalTitle = '', modalItem = {}, ...exProps} = props;
+
+    const [showModal, setShowModal] = React.useState(false);
```
**Можно лучше**

Если интересно, можно создать **кастомный хук** для управления модальным окном:

```js
import { useState, useCallback } from "react";

// кастомные хуки всегда должны начинаться с глагола `use`, чтобы реакт понял, что это хук. Он следит за их вызовами
export const useModal = () => {
  const [isModalOpen, setIsModalOpen] = useState(false);

  // `useCallback` нужен для того, чтобы зафиксировать ссылку на функцию. Таким образом уменьшится кол-во перерисовок компонента, куда будет передана эта функция
  const openModal = useCallback(() => {
    setIsModalOpen(true);
  }, []);

  const closeModal = useCallback(() => {
    setIsModalOpen(false);
  }, []);

  return {
    isModalOpen,
    openModal,
    closeModal,
  };
};

```

Этот код помещают в отдельный файл `useModal.js` в папке `hooks` и импортируют функцию туда, где нужно создать показатели открытости/закрытости попапа (у каждого попапа они свои)

И теперь нужно просто вызывать хук в том компоненте, где нужно управлять попапом. Всего одна строчка кода будет:

```js
  const { isModalOpen, openModal, closeModal } = useModal();

````

gennady-bars commented on src/components/hocs/withModal.jsx:
```diff
@@ -0,0 +1,46 @@
+        <>
+            <WrappedComponent {...exProps} onClick={ openModal } />
+            {showModal  && 
+                <Modal type={modalType} title={modalTitle} item={modalItem} Close={ closeModal } >
```
`Close`


Любые переменные в проекте нужно называть в `camelCase` по чек-листу. Это значит, что они должны быть с **маленькой буквы**, если это не классы или компоненты реакта

gennady-bars commented on src/components/app/app.jsx:
```diff
@@ -1,17 +1,59 @@
+          setState({...state, data: data.data, isLoading: false})  
+      })
+      } else {
+        const newState = {...state, data:[], isLoading: false, errorText: res.status + " " + res.statusText  }
```
Не нужно дублировать обработки ошибки тут, так как для этого есть блок `catch`


просто тут нужно выкидывать ошибку


Вот так выглядит проверка на `ok`


```js
    fetch(API_URL)
      .then((res) => {
        if (res.ok) {  // проверили
          return res.json();  // распарисили ответ
        }
        return Promise.reject(`Ошибка ${res.status}`);  // выкинули ошибку, если не `ok`
      })
      .then((data) => {  // тут пришли готовые данные
        setState({ ...state, data: data.data, isLoading: false });
      })
      .catch((err) => {
```

gennady-bars commented on src/components/burger-ingredients/ingredients/ingredients.jsx:
```diff
@@ -1,50 +1,22 @@
+import withModal from "../../hocs/withModal";
 import styles from './ingredients.module.css';
 
+const WithModalCard = withModal(IngredientCard) 
```
**Можно лучше**

 Лучше не делать в **каждом** ингредиенте по модальному окну. Нужно было сделать одно окно для деталей в ` BurgerIngredients` (или `App`) и передавать в каждый ингредиент функцию **клика**, в которой бы назначались данные для показа в окне. Это поможет в спринте, где **роутинг** будете изучать и реализовывать

gennady-bars commented on src/components/burger-ingredients/ingredient-card/ingredient-card.jsx:
```diff
@@ -0,0 +1,24 @@
+
+function IngredientCard({item, onClick}) {
+    return (
+        <>
```
`<>` пустые теги нужны только тогда, когда нет **общего** тега (компонента), который оборачивает всю верстку

gennady-bars commented on src/components/burger-ingredients/ingredient-card/ingredient-card.jsx:
```diff
@@ -0,0 +1,24 @@
+function IngredientCard({item, onClick}) {
+    return (
+        <>
+            <li className={styles.li} onClick= {onClick} >
```
Вам нужно **отформатировать** код, чтобы были видны нормальные отступы и вложенность одних компонентов в другие. И `js`-код тоже будет удобно читать. Это можно сделать комбинацией клавиш `SHIFT + ALT + F` или с помощью `prettier`



```jsx

  return (
    <li className={styles.li} onClick={onClick}>
      <Counter count={Math.floor(Math.random() * 100)} size="default" />
      <img className="ml-4 mr-4" src={item.image} alt={item.name} />
      <p className={`mt-1 mb-1 text text_type_digits-default ${styles.price}`}>
        {item.price} <CurrencyIcon type="primary" />
      </p>
      <p
        className={`text text text_type_main-default ${styles["text-center"]}`}
      >
        {item.name}
      </p>
    </li>
  );

```

gennady-bars commented on src/components/burger-ingredients/ingredient-card/ingredient-card.jsx:
```diff
@@ -0,0 +1,24 @@
+}
+
+IngredientCard.propTypes = {
+    item: typeOfIngredient
```
Еще `onClick` забыли типизировать тут

gennady-bars commented on src/components/modal/modal.jsx:
```diff
@@ -0,0 +1,59 @@
+  const modalRoot = document.getElementById('modal-root');  
+  
+  const handleKeyDown = (e) => {
+    if(e.keyCode === 27) {
```
Цифры лучше не использовать для определения клавиш, так как они не говорят ничего о себе (какая это клавиша). Лучше использовать строки. `(evt.key === “Escape”)`. И в чек-листе есть пункт про то, что нельзя использовать `магические` значения, которые не являются понятными любому разработчику.

gennady-bars commented on src/components/modal/modal.jsx:
```diff
@@ -0,0 +1,59 @@
+
+function Modal(props) {
+
+  const modalRoot = document.getElementById('modal-root');  
```
Этот блок достаточно 1 раз найти на 7й строчке, чтобы не делать этого при каждой перерисовке компонента

gennady-bars commented on src/components/modal/modal.jsx:
```diff
@@ -0,0 +1,59 @@
+  React.useEffect( () => {
+    window.addEventListener('keydown', handleKeyDown);
+    return () => {
+        window.removeEventListener('keydown', handleKeyDown);
```
Правильно удаляется обработчик

gennady-bars commented on src/components/modal/modal.jsx:
```diff
@@ -0,0 +1,59 @@
+
+  const modalRoot = document.getElementById('modal-root');  
+  
+  const handleKeyDown = (e) => {
```
Обработчик `Escape` нужно поместить внутрь `useEffect` , чтобы не менялась ссылка на функцию при перерисовке компонента.

gennady-bars commented on src/components/modal/modal.jsx:
```diff
@@ -0,0 +1,59 @@
+          <div className={`pt-10 pb-10 pr-10 pl-10 ${styles.modal}`}>
+              <div className={styles.head}>
+                  <span className="text text_type_main-large mr-6">{props.title}</span>
+                  <span className={`text text_type_main-medium ${styles.close}`} onClick={props.Close}>✖</span>
```
Для крестика есть компонент `CloseIcon` в яндекс-библиотеке

gennady-bars commented on src/components/modal/modal.jsx:
```diff
@@ -0,0 +1,59 @@
+  error: 'Непредвиденная ошибка'
+}
+
+Error.propTypes = {
```
Эта типизация не нужна тут. Вы не типизируете компонент `Error`,  так как его тут не существует

gennady-bars commented on src/components/app/app.jsx:
```diff
@@ -1,17 +1,59 @@
+    setState({...state, data: [], isLoading: true})
+    fetch(API_URL) 
+    .then( res => { 
+      if(res.ok) {
```
Отлично, что проверяете ответ на `ok`

gennady-bars commented on src/components/app/app.jsx:
```diff
@@ -1,17 +1,59 @@
+        setState(newState)
+      }
+    })
+    .catch( err =>  {
```
Отлично, что ловите возможные ошибки в конце каждого запроса к серверу

gennady-bars commented on src/components/app/app.jsx:
```diff
@@ -1,17 +1,59 @@
+
+  useEffect(() => {
+    fetchData()
+  }, [])
```
Отлично, что используете пустой массив зависимостей `[]`, чтобы `useEffect` сработал 1 раз

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:



* Отлично, что проверяете ответ на `ok`
* Отлично, что ловите возможные ошибки в конце каждого запроса к серверу
* Отлично, что используете пустой массив зависимостей `[]`, чтобы `useEffect` сработал 1 раз



но есть некоторые недочеты:



*  При открытии попапа в консоли ошибка https://disk.yandex.ru/i/2-RBTDGoKJtTPQ
* Любые переменные в проекте нужно называть в `camelCase` по чек-листу. Это значит, что они должны быть с **маленькой буквы**, если это не классы или компоненты реакта
* `<>` пустые теги нужны только тогда, когда нет **общего** тега (компонента), который оборачивает всю верстку
* Вам нужно **отформатировать** код, чтобы были видны нормальные отступы и вложенность одних компонентов в другие. И `js`-код тоже будет удобно читать. Это можно сделать комбинацией клавиш `SHIFT + ALT + F` или с помощью `prettier`
* Цифры лучше не использовать для определения клавиш, так как они не говорят ничего о себе (какая это клавиша). Лучше использовать строки. `(evt.key === “Escape”)`. И в чек-листе есть пункт про то, что нельзя использовать `магические` значения, которые не являются понятными любому разработчику.





**Можно лучше**

* Лучше не делать в **каждом** ингредиенте по модальному окну. Нужно было сделать одно окно для деталей в ` BurgerIngredients` (или `App`) и передавать в каждый ингредиент функцию **клика**, в которой бы назначались данные для показа в окне. Это поможет в спринте, где **роутинг** будете изучать и реализовывать



Исправьте, пожалуйста, недочеты и работа будет принята. Пожалуйста, проверьте **работоспособность проекта** и наличие возможных ошибок в консоли браузера (кнопка `F12`) перед отправкой на ревью. 

Напоминаю, что работа может быть принята только после исправления всех критических замечаний `Нужно исправить`.

Удачного рефакторинга кода.

andrey340 commented:
Благодарю за ОС.
Исправил ошибки и произвел рефакторинг функционала модальных окон.
С уважением Волобуев Андрей


Thanks for the feedback.
Fixed bugs and refactored the functionality of modals.
With regards, Andrey Volobuev

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
